lib/Dialect/TritonNvidiaGPU/Transforms/InterleaveTMem.cpp

```diff
@@ -229,6 +233,68 @@ bool trySinkOp(Operation *op, Value buffer) {
   return sinkOps(buffer, useChain);
 }
 
+// Hoist tmem_stores as close to their operand def as possible to reduce
+// register pressure.
+bool hoistOps(Value buffer, ArrayRef<Operation *> defChain,
+              ArrayRef<OpOperand> storeOperands) {
+  Operation *insertAfter = nullptr;
+  Operation *prev = defChain.back()->getPrevNode();
+  while (prev) {
+    insertAfter = prev;
+    bool dep = false;
+    for (auto operand : getNestedOperands(defChain.back())) {
+      if (llvm::is_contained(prev->getResults(), operand)) {
+        dep = true;
+        break;
+      }
+    }
+    for (const OpOperand &operand : storeOperands) {
+      if (llvm::is_contained(prev->getResults(), operand.get())) {
+        dep = true;
+        break;
+      }
+    }
+    // Don't hoist past barrier signals, since they may guard the liverange
+    // of the buffer.
+    if (opAlias(prev, buffer))
```

Comment:
The alias check for hoisting stores needs to be different than for sinking loads. It should check to not hoist above `ttng.wait_barrier` and also needs to look for aliasing read effects